icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1239,147 +1563,54 @@ public static String canonicalize(String localeID){
-                        ULocale replaceLocale = new ULocale(languageAlias.get(
-                            (new StringBuilder(parser.getLanguage().length() + 1 + parser.getVariant().length()))
-                                .append(parser.getLanguage())
-                                .append("_")
-                                .append(variant)
-                                .toString())
-                            .get("replacement")
-                            .getString());
-                        StringBuilder replacedVariant = new StringBuilder(parser.getVariant().length());
-                        for (String current : variants) {
-                            if (current.equals(variant)) continue;
-                            if (replacedVariant.length() > 0) replacedVariant.append("_");
-                            replacedVariant.append(current);
-                        }
-                        parser = new LocaleIDParser(
-                            (new StringBuilder(localeID.length()))
-                                .append(lscvToID(replaceLocale.getLanguage(),
-                                    !parser.getScript().isEmpty() ? parser.getScript() : replaceLocale.getScript(),
-                                    !parser.getCountry().isEmpty() ? parser.getCountry() : replaceLocale.getCountry(),
-                                    replacedVariant.toString()))
-                                .append(parser.getName().substring(parser.getBaseName().length()))
-                                .toString());
-                    } catch (MissingResourceException e) {
-                    }
-                }
-            }
-
-            // language _ script _ country
-            // ug_Arab_CN -> ug_CN
-            if (!parser.getScript().isEmpty() && !parser.getCountry().isEmpty()) {
-                try {
-                    ULocale replaceLocale = new ULocale(languageAlias.get(
-                        lscvToID(parser.getLanguage(), parser.getScript(), parser.getCountry(), null))
-                        .get("replacement")
-                        .getString());
-                    parser = new LocaleIDParser((new StringBuilder(localeID.length()))
-                        .append(lscvToID(replaceLocale.getLanguage(),
-                            replaceLocale.getScript(),
-                            replaceLocale.getCountry(),
-                            parser.getVariant()))
-                        .append(parser.getName().substring(parser.getBaseName().length()))
-                        .toString());
-                } catch (MissingResourceException e) {
-                }
+        boolean knownCanonicalized = false;
+        String name = parser.getName();
+        if (!isKnownCanonicalizedLocale(name)) {
+            AliasReplacer replacer = new AliasReplacer(
+                parser.getLanguage(), parser.getScript(), parser.getCountry(), parser.getVariant().toLowerCase(),
```

Comment:
not `toLowerCase()` -- that does more work than we need for ASCII, and with the implicit default locale, it will fail on Turkish etc. systems

I feel sure that we have one or more helper classes for ASCII lowercasing.

Reply:
good catch AsciiUtil.toLowerString() should do it.